docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+
+The JobSpec contains a single pod template.  When the job controller creates a particular
+pod, it copies the pod template and modifies it in some way to make that pod distinctive.
+Whatever is distinctive about that pod is its *identity*.
+
+We consider several options.
+
+#### Index Substitution Only
+
+The job controller substitutes only the *completion index* of the pod into the
+pod template when creating it.  The JSON it POSTs differs only in a single
+fields.  This field could be a label, an annotation, or an environment
+variable.
+
+This option makes no changes to the JobSpec.
+
+**TODO**: explore why annotation, as opposed to label or env var.
+
+#### Multiple Substitution
+
+In this option, the JobSpec is extended to include a list of values to substitute,
+and which fields to substitute them into.  For example, a worklist like this:
+
+```
+FRUIT_COLOR=green process-fruit -a -b -c -f apple.txt --remove-seeds
+FRUIT_COLOR=yellow process-fruit -a -b -c -f banana.txt
+FRUIT_COLOR=red process-fruit -a -b -c -f cherry.txt --remove-pit
+```
+
+Can be broken down into a template like this, with three parameters
+
+```
+<custom env var 1>; process-fruit -a -b -c <custom arg 1> <custom arg 1>
+```
+
+and a list of parameter tuples, like this:
+
+```
+("FRUIT_COLOR=green", "-f apple.txt", "--remove-seeds")
+("FRUIT_COLOR=yellow", "-f banana.txt", "")
+("FRUIT_COLOR=red", "-f cherry.txt", "--remove-pit")
+```
+
+The JobSpec can be extended to hold a list of parameter tuples (which
+are more easily expressed as a list of lists of individual parameters).
+For example:
+
+```
+apiVersion: extensions/v1beta1
+kind: Job
+...
+spec:
+  completions: 3
+  ...
+  template:
+    ...
+  perCompletionArgs:
+    container: 0
+      -
+        - "-f apple.txt"
+        - "-f banana.txt"
+        - "-f cherry.txt"
+      -
+        - "--remove-seeds"
+        - ""
+        - "--remove-pit"
+  perCompletionEnvVars:
+    - name: "FRUIT_COLOR"
+      - "green"
+      - "yellow"
+      - "red"
+```
+
+However, just providing custom env vars, and not arguments, is sufficient
+for many use cases: parameter can be put into env vars, and then
+substituted on the command line.
+
+#### Comparison
+
+The multiple substitution approach:
+
+- keeps the *per completion parameters* in the JobSpec.
+- Drawback: makes the job spec large for job with thousands of completions. (But for very large jobs, the work-queue style or another type of controller, such as  map-reduce or spark, may be a better fit.)
+
+The index-only approach:
+
+- requires that the user keep the *per completion parameters* in a separate storage, such as a configData or networked storage.
+- Drawback: while in separate storage, they could be mutatated, which would have unexpected effects
+- Drawback: Logic for using index to lookup paramters needs to be in the Pod.
+- Drawback: CLIs and UIs are limited to using the "index" as the identity of a pod
```

Comment:
Whether the per-completion data is in a pod or in ConfigMap, annotations could record the source of the data.